### PR TITLE
refactor inbox storage

### PR DESCRIPTION
**Pull request by mmaxim**

### CODE REVIEW

File: go/chat/storage/inbox.go [hunk 1]
```diff
@@ -64,19 +59,43 @@ func (q inboxDiskQuery) match(other inboxDiskQuery) bool {
...
-	InboxVersion  chat1.InboxVers            `codec:"I"`
-	Conversations []types.RemoteConversation `codec:"C"`
-	Queries       []inboxDiskQuery           `codec:"Q"`
+type inboxDiskIndex struct {
+	ConversationIDs []chat1.ConversationID `codec:"C"`
+	Queries         []inboxDiskQuery       `codec:"Q"`
+}
+
+func (i *inboxDiskIndex) mergeConvs(convIDs []chat1.ConversationID) {
+	m := make(map[string]chat1.ConversationID)
```
Comment (joshblum) [inline go/chat/storage/inbox.go]: prealloc?

File: go/chat/storage/inbox.go [hunk 2]
```diff
@@ -172,130 +193,150 @@ func (i *Inbox) readDiskInbox(ctx context.Context, uid gregor1.UID, useInMemory
...
-	i.summarizeConvs(ibox.Conversations)
-	i.Debug(ctx, "writeDiskInbox: uid: %s version: %d disk version: %d server version: %d convs: %d",
-		uid, ibox.InboxVersion, ibox.Version, ibox.ServerVersion, len(ibox.Conversations))
-	inboxMemCache.Put(uid, &ibox)
-	switch i.flushMode {
-	case InboxFlushModeActive:
-		return i.flushLocked(ctx, uid)
-	case InboxFlushModeDelegate:
-		return nil
+	return convs[0], nil
```
Comment (joshblum) [inline go/chat/storage/inbox.go]: can you sanity check the length before indexing? i know the other method returns a `MissError`, but to be future safe

File: go/chat/storage/inbox.go [hunk 3]
```diff
@@ -599,49 +601,71 @@ func (i *Inbox) Read(ctx context.Context, uid gregor1.UID, query *chat1.GetInbox
...
-	ierr := i.G().LocalChatDb.Delete(i.dbKey(uid))
-	if ierr != nil {
+	iboxIndex, err := i.readDiskIndex(ctx, uid, true)
+	if err != nil {
+		i.Debug(ctx, "Clear: failed to read index: %s", err)
+		return err
+	}
+	for _, convID := range iboxIndex.ConversationIDs {
+		if ierr := i.G().LocalChatDb.Delete(i.dbConvKey(uid, convID)); ierr != nil {
+			return NewInternalError(ctx, i.DebugLabeler,
```
Comment (joshblum) [inline go/chat/storage/inbox.go]: just log here, no need to kill whole operation (same below)

File: go/chat/storage/inbox.go [hunk 4]
```diff
@@ -699,76 +722,75 @@ func (i *Inbox) NewConversation(ctx context.Context, uid gregor1.UID, vers chat1
...
 	tlfID, err := chat1.TeamIDToTLFID(teamID)
 	if err != nil {
 		i.Debug(ctx, "getConvsForTeam: teamIDToTLFID failed: %v", err)
 		return nil
 	}
-	for i := range convs {
-		conv := &convs[i]
-		if conv.Conv.GetMembersType() == chat1.ConversationMembersType_TEAM && conv.Conv.Metadata.IdTriple.Tlfid.Eq(tlfID) {
+	for _, convID := range index.ConversationIDs {
+		conv, err := i.readConv(ctx, uid, convID)
```
Comment (joshblum) [inline go/chat/storage/inbox.go]: bummer we have to read everything here, how often is this method called? maybe we could keep an index for convs by tlf id too?

Comment (mmaxim) [reply go/chat/storage/inbox.go]: Very rarely, only if the team wide retention is set.

**Pull request merged**